Первая часть финального проекта

DrEremin opened the pull request:
Задание со звездочкой постараюсь выполнить вместе со 2 частью проекта

DrEremin committed:
Add all classes for getBalance(), putMoney(), takeMoney().
Add package controllers: class InternetBankController;
Add package models: class BankAccount;
Add package repositories: interface BankAccountRepository;
Add package services: class BankAccountService;
Add package dto: classes: BalanceDTO, StatusOperationDTO
and interface BankAccountDTO;
Add package impl to package dto: classes: UserIdDTO and
UserAndMoneyDTO.
Fill in the application.properties file.

DrEremin committed:
Add exception handling, add/remove accounts.
Add class ExceptionController to package controllers.
Create package exceptions with classes:
DataMissingException, IncorrectNumberException,
NotEnoughMoneyException, UniquenessViolationException.
Add methods to the class InternetBankController and
InternetBankService: createAccount(), deleteAccount.

DrEremin committed:
Add database damp and README.md

DrEremin committed:
Add model and repository for entity operation

Add class ru.dreremin.internerbank.models.Operation;
Add interface
ru.dreremin.internetbank.repositories.OperationRepository;
Add screenshot of data logic model.

DrEremin committed:
Add class OperationService

Replace the interface BankAccountDTO with an abstract class;
Add new fields of following types to BankAccountDTO:
LocalDate, LocalTime, ZoneId;
Add the saveOperation() method to class OperationService;
Add the ability to save the operation to the 'operation' table
in the ServiceAccount.getBalance() method.

DrEremin committed:
Add handlers related with date, time & time zone

Add handler of the class DateTimeParseException;
Add handler of the class DateTimeException;
Add handler of the class ZoneRulesException.

DrEremin committed:
Add the save of operations to BankAccountService

Add the OperationService.save() method to
BankAccountService.getMoney() and BankAccountService.putMoney().

DrEremin committed:
Add transactions to BankAccountService

Add @Transaction annotation to BankAccountService.getBalance(),
BankAccountService.putMoney(), BankAccountService.takeMoney();
Update DLM to README.md.

DrEremin committed:
Add class UserIdAndMoneyAndRecipientIdDTO

DrEremin committed:
Add cl. TransferRecipient & TransferRecipientPkey

Add ru.dreremin.internetbank.models.TransferRecipient and
ru.dreremin.internetbank.models.TransferRecipientPkey classes;
Delete lombok in whole project;
Add getters and constructors in whole project;
Fix imports in whole project.

DrEremin committed:
Add new classes and method transferMoney()

Add ru.dreremin.internetbank.repositories.TransferRecipientRepository,
ru.dreremin.internetbank.services.TransferRecipientService;
Add ru.dreremin.internetbank.services.BankAccountService.transferMoney();
Make BankAccountRepository the heir of JpaRepository.

DrEremin committed:
Add BankAccountController.transferMoney()

Add ru.dreremin.internetbank.controllers.BankAccountController
.transferMoney();
Add ru.dreremin.internetbank.exceptions.SameIdException and handler
for it to ru.dreremin.internetbank.controllers.ExceptionController;
Add SameIdException to ru.dreremin.internetbank.dto.BankAccountDTO
.validation().

DrEremin committed:
Add enum OperationTypes and renames

Rename field ru.dreremin.internetbank.dto.BankAccountDTO
    .userId to clientId;
Rename class ru.dreremin.internetbank.dto.impl.UserIdDTO to ClientIdDTO;
Rename class ru.dreremin.internetbank.dto.impl
    .UserIdAndMoneyAndRecipientIdDTO to SenderIdAndMoneyAndRecipientIdDTO;
Rename method ru.dreremin.internetbank.repositories
    .BankAccountRepository.getBankAccountByUserId() to
    getBankAccountByClientId();
Rename class ru.dreremin.internetbank.dto.impl.UserIdAndMoneyDTO
    to ClientIdAndMoneyDTO.

DrEremin committed:
Add functional for the operation getOperationList

Add classes ru.dreremin.internetbank.models.OperationDescription,
ru.dreremin.internetbank.repositories.OperationDescriptionRepository,
ru.dreremin.internetbank.services.OperationDescriptionService,
ru.dreremin.internetbank.dto.DateTimesOfPeriodWithZoneIdDTO,
ru.dreremin.internetbank.exceptions.DateTimeOutOfBoundsException
and method
ExceptionsController.handlerDateTimeOutOfBoundsException().

DrEremin committed:
Add OperationListDTO

Add ru.dreremin.internetbank.dto.OperationListDTO;
Add method ru.dreremin.internetbank.controllers.BankAccountController
    .getOperationList();
Add @Transactional to all repositories.

DrEremin committed:
Add method getAllWithinRange()

Add method ru.dreremin.internetbank.repositories
    .OperationDescriptionRepository.getAllWithinRange().

DrEremin committed:
Add handling nullable dates in getOperationList()

Add ru.dreremin.internetbank.dto.DateTimesOfPeriodWithZoneIdDTO
    .isNullableTimePoint and corresponding getter;
Add tables of database to README.md;
Add new the dump of database.

DrEremin committed:
Add cl.: Client, ClientRepository, ClientService

Add ru.dreremin.internetbank.models.Client,
Add ru.dreremin.internetbank.repositories.ClientRepository,
Add ru.dreremin.internetbank.services.ClientService.

DrEremin committed:
Add unit-tests. Set properties profiles

Add unit-test classes:
ru.dreremin.internetbank.services.OperationServiceTest,
ru.dreremin.internetbank.services.TransferRecipientServiceTest;
Add new files: application-main.properties,
application-test.properties.

DrEremin committed:
Add unit-test OperationDescriptionServiceTest

Add class
ru.dreremin.internetbank.services.OperationDescriptionServiceTest.

DrEremin committed:
Add unit-test BankAccountServiceTest

Add class ru.dreremin.internetbank.services.BankAccountServiceTest.
Test the following methods of class
ru.dreremin.internetbank.services.BankAccountService:
getBalance(), putMoney(), takeMoney().

DrEremin committed:
Revision of the architecture

Reduce coupling  in the InternetBankController and BankAccountService:
* Create ManagementInternetBankController.
  Methods deleteAccount() and createAccount() from
  InternetBankController have been transferred to it;
* Create OperationsController.
  Method getOperationList() from InternetBankController have been
  transferred to it.
* Create service BankAccountManager.
  Methods deleteAccount() and createAccount() from BankAccountService
  have been transferred to it.

DrEremin committed:
Full to remake the class BankAccountServiceTest

DrEremin committed:
Add ClientServiceTest and BankAccountManagerTest

vrnsky commented on src/main/java/ru/dreremin/internetbank/ServletInitializer.java:
```diff
@@ -3,7 +3,7 @@
 import org.springframework.boot.web.servlet.support.SpringBootServletInitializer;
 
-public class ServletInitializer extends SpringBootServletInitializer {
+    public class ServletInitializer extends SpringBootServletInitializer {
```
Лишний пробел

vrnsky commented on src/main/java/ru/dreremin/internetbank/controllers/InternetBankController.java:
```diff
@@ -0,0 +1,67 @@
+            throws IncorrectNumberException, DataMissingException,
+            NotEnoughMoneyException {
+
+        dto.validation();
```
Тут лучше заиспользовать библиотеки валидации, можно посмотреть в сторону Hibernate Validation, Spring Validation

vrnsky commented on src/main/java/ru/dreremin/internetbank/controllers/InternetBankController.java:
```diff
@@ -0,0 +1,67 @@
+
+        dto.validation();
+        bankAccountService.putMoney(dto);
+        log.info("Put money operation was completed successfully");
```
Плохой подход к написанию логов, так как по данному логу невозможно будет точно идентифицировать кому были начислены деньгие

vrnsky commented on src/main/java/ru/dreremin/internetbank/controllers/InternetBankController.java:
```diff
@@ -0,0 +1,67 @@
+    }
+
+    @PatchMapping(value="/put-money", consumes="application/json")
+    public StatusOperationDTO putMoney(@RequestBody ClientIdAndMoneyDTO dto)
```
Нарушение принципала принципа единственной отвественности в названии дто, лучше переименовать

vrnsky commented on src/main/java/ru/dreremin/internetbank/controllers/ManagementBankAccountController.java:
```diff
@@ -0,0 +1,44 @@
+@RequestMapping("/account-management")
+public class ManagementBankAccountController {
+
+    private final BankAccountManager manager;
```
Spring Boot и Lombok можно подружить, чтобы не писать конструкторы - достаточно добавить аннотацию над классом `@RequiredArgsConstructor`

vrnsky commented on src/main/java/ru/dreremin/internetbank/dto/DateTimesOfPeriodWithZoneIdDTO.java:
```diff
@@ -0,0 +1,81 @@
+{
+    "startDate" : "yyyy-mm-dd",
+    "startTime" : "hh-mm-ss",
+    "endDate" : "yyyy-mm-dd",
```
Тут такого быть не должно, закомментированный код/JSON лучше обрать

vrnsky commented on src/main/java/ru/dreremin/internetbank/models/Operation.java:
```diff
@@ -0,0 +1,109 @@
+    public int getOperationTypeId() { return operationTypeId; }
+
+    public ZonedDateTime getTimeAndDateOfOperation() {
+        return timeAndDateOfOperation;
```
and тут лишнее `timeDateOfOperation` выглядит более хорошо

vrnsky commented on src/main/java/ru/dreremin/internetbank/services/BankAccountManager.java:
```diff
@@ -0,0 +1,70 @@
+public class BankAccountManager {
+
+    ClientService service;
+    BankAccountRepository repository;
```
Почему не `private final...`?

vrnsky commented on src/main/java/ru/dreremin/internetbank/services/BankAccountService.java:
```diff
@@ -0,0 +1,187 @@
+                        clientIdAndMoneyDTO.getClientId());
+
+        if (optionalBankAccount.isEmpty()) {
+            String message = "Client with this id does not exist";
```
Дважды повторяется строчка - лучше вынести в константу

vrnsky commented on src/main/java/ru/dreremin/internetbank/services/ClientService.java:
```diff
@@ -0,0 +1,32 @@
+    }
+
+    public boolean isClientWithThisIdExist(long id) {
+
```
Лишняя пустая строки

vrnsky commented on src/main/resources/application-main.properties:
```diff
@@ -0,0 +1,14 @@
+spring.datasource.url=jdbc:postgresql://localhost:5432/internet_bank
+spring.datasource.username=dr_eremin
```
Тут лучше не светить credentials'ами, факультативно советую посмотреть про такой инструмент, как Vault

vrnsky commented on src/test/java/ru/dreremin/internetbank/services/TransferRecipientServiceTest.java:
```diff
@@ -0,0 +1,63 @@
+
+        verify(repository, times(1)).save(captor.capture());
+
+        assertEquals(standard, captor.getValue());
```
Тесты выглядят хорошо, но если хочется сделать более профессионально, то посмотрите в сторону TestContainers

DrEremin committed:
Add ManagementBankAccountController

Add test class ru.dreremin.internetbank.controllers
    .ManagementBankAccountController;
Add util package with c classes:
CustomLocalDateSerializer, CustomLocalTimeSerializer,
CustomZoneIdSerializer, JsonConverter.

DrEremin committed:
Do code refactoring

Order imports;
Replace constructors and getters with the lombok library where possible;
Rename some classes in dto package.
